### ISSUE TITLE
Owner invalidation invalidates Pointer to Owner (instead of only "into Owner")

### ISSUE THREAD

**Issue by brevzin** [mgehre/llvm-project#66]
This program:

```cpp
#include <string>

void trim(std::string& str)
{
    auto start = str.find_first_not_of(' ');
    str.erase(0, start);
    auto stop  = str.find_last_not_of(' ');
    str.erase(stop+1);
}
```

emits:

```
/home/brevzin/sandbox/foo.cxx:7:18: warning: dereferencing a dangling pointer [-Wlifetime]
    auto stop  = str.find_last_not_of(' ');
                 ^~~
/home/brevzin/sandbox/foo.cxx:6:5: note: modified here
    str.erase(0, start);
    ^~~
/home/brevzin/sandbox/foo.cxx:8:5: warning: dereferencing a dangling pointer [-Wlifetime]
    str.erase(stop+1);
    ^~~
/home/brevzin/sandbox/foo.cxx:6:5: note: modified here
    str.erase(0, start);
    ^~~
```

There's no dangling here, this is fine. 

Here's a creduce-assisted repro:

```cpp
namespace std {
inline namespace {
class basic_string {
public:
  void a();
};
}
}
void b(std::basic_string &c) {
  c.a();
  c;
}
```

which emits:

```
foo_pp.cxx:11:3: warning: dereferencing a dangling pointer [-Wlifetime]
  c;
  ^
foo_pp.cxx:10:3: note: modified here
  c.a();
  ^
```

**Comment by mgehre** [comment]
Thanks for reporting!

The call to non-const `a` is supposed to invalidate Pointer (like iterators) that point *into* the std::string. Seems we currently do too much here, and also invalidate Pointers that point *at* the std::string (like the reference in your example).

**Comment by mgehre** [comment]
I tried to highlight the issue in the following examples:
https://godbolt.org/z/klPC0r
